### ISSUE TITLE
How can I implement PySwarms in regression problems?

### ISSUE THREAD

**Issue by JaveJao** [ljvmiranda921/pyswarms#42]
I actually have one, sir! I am trying to compare the implementation of BP NN vs PSO NN. I have a code for BP NN using TensorFlow. Here is my code:

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from tensorflow.contrib import learn
from sklearn.metrics import accuracy_score
from sklearn.pipeline import Pipeline
from sklearn import datasets, linear_model
from sklearn import model_selection
from sklearn import preprocessing

import tensorflow as tf
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import math
import calendar
import time

region = "NCR"
filename = region + ".csv"
x = np.array(pd.read_csv(filename, usecols=[1,2,3,4], header=None))
y = np.array((pd.read_csv(filename, usecols=[5], header=None)))
normalized_X = preprocessing.scale(x)

X_train, X_test, Y_train, Y_test = model_selection.train_test_split(
normalized_X, y, test_size=0.33333333, shuffle=False, stratify=None)

total_len = X_train.shape[0]

# Parameters
learning_rate = 0.01
training_epochs = 600
batch_size = X_train.shape[0]
display_step = 1
model_path = "/tmp/models/" + region + ".pkt"

# Network Parameters
n_hidden_1 = 256 
n_hidden_2 = 256
n_hidden_3 = 256
n_hidden_4 = 256
n_input = X_train.shape[1]
n_classes = 1

# TensorFlow graph input
normalized_X = tf.placeholder("float", [None, 4])
y = tf.placeholder("float", [None, 1])

# Create model
def multilayer_perceptron(normalized_X, weights, biases):
    # Hidden layer with RELU activation
    layer_1 = tf.add(tf.matmul(normalized_X, weights['w1']), biases['b1'])
    layer_1 = tf.nn.relu(layer_1)
    
    # Hidden layer with RELU activation
    layer_2 = tf.add(tf.matmul(layer_1, weights['w2']), biases['b2'])
    layer_2 = tf.nn.relu(layer_2)

    # Hidden layer with RELU activation
    layer_3 = tf.add(tf.matmul(layer_2, weights['w3']), biases['b3'])
    layer_3 = tf.nn.relu(layer_3)

    # Hidden layer with RELU activation
    layer_4 = tf.add(tf.matmul(layer_3, weights['w4']), biases['b4'])
    layer_4 = tf.nn.relu(layer_4)
    
    # Output layer with linear activation
    out_layer = tf.matmul(layer_4, weights['output']) + biases['output']
    return out_layer

# Define weights and bias
weights = {
    'w1': tf.Variable(tf.random_normal([n_input, n_hidden_1], 0, 1)),
    'w2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2], 0, 1)),
    'w3': tf.Variable(tf.random_normal([n_hidden_2, n_hidden_3], 0, 1)),
    'w4': tf.Variable(tf.random_normal([n_hidden_3, n_hidden_4], 0, 1)),
    'output': tf.Variable(tf.random_normal([n_hidden_4, n_classes], 0, 1))
}
biases = {
    'b1': tf.Variable(tf.random_normal([n_hidden_1], 0, 1)),
    'b2': tf.Variable(tf.random_normal([n_hidden_2], 0, 1)),
    'b3': tf.Variable(tf.random_normal([n_hidden_3], 0, 1)),
    'b4': tf.Variable(tf.random_normal([n_hidden_4], 0, 1)),
    'output': tf.Variable(tf.random_normal([n_classes], 0, 1))
}

# Construct model
pred = multilayer_perceptron(normalized_X, weights, biases)

# Define cost and optimizer functions
cost = tf.reduce_mean(tf.square(pred-y))
optimizer = tf.train.AdamOptimizer(learning_rate=learning_rate).minimize(cost)

# Saver op to save and restore all the variables
saver = tf.train.Saver()

# Launch the training session
start_time = time.time()
print("Initializing training session...\n")
with tf.Session() as sess:
    init = tf.global_variables_initializer()
    sess.run(init)

    # Training cycle
    for epoch in range(training_epochs):
        avg_cost = 0.0
        total_batch = int(total_len/batch_size)
        
        # Loop over all batches
        for i in range(total_batch):
            batch_x = X_train[i*batch_size:(i+1)*batch_size]
            batch_y = Y_train[i*batch_size:(i+1)*batch_size]
            feed_dict = {
                normalized_X: batch_x,
                y: batch_y,
            }
            
            # Run optimization op (backprop) and cost op (to get loss value)
            _, c, p = sess.run([optimizer, cost, pred], feed_dict=feed_dict)
            
            # Compute average cost
            avg_cost += c / total_batch

        # Sample prediction
        actual_value = batch_y
        predicted_value = p
        
        # Display outputs per epoch
        if epoch % display_step == 0:
            # print ("Epoch:", '%03d' % (epoch+1), "cost:", "{:.3f}".format(avg_cost))
            print ("Epoch:", '%03d' % (epoch+1))
            print ("----------------------------------------------------------------------------------------------------------------")
            j = 1
            year = 2008
            for i in range(len(actual_value)):
                if i % 12 == 0:
                    print ("\n" + str(year))
                    j = 1
                    year += 1
                else:
                    j += 1
                print (calendar.month_name[j] + ":", "Actual value =", '%01f' % actual_value[i] + " mwh, Predicted value =", '%01f' %  predicted_value[i] + \
                " mwh, Forecast error:", "{:.6f}".format(np.mean(np.abs((actual_value[i] - predicted_value[i]) / actual_value[i])) * 100))
                
            print ("================================================================================================================")

    print ("Training Finished!\n")
    print("Execution Time:", (time.time() - start_time), "seconds")
    
    # Save model weights
    save_path = saver.save(sess, model_path)
    
    # Test model
    predictions = sess.run(pred, feed_dict={normalized_X: X_test})
    
    # Accuracy metrics
    MAPE = 1/len(actual_value) * np.sum(abs(actual_value-predicted_value)/actual_value) * 100
    Accuracy = 100 - MAPE
    print("MAPE:", "{:.6f}".format(MAPE), "\nPrediction Accuracy:", "{:.6f}".format(Accuracy) + "%")
    
    print ("\nPredictions:\n",predictions)

**Comment by ljvmiranda921** [comment]
Hmmm, I am not sure how we can incorporate a different optimization scheme in Tensorflow because it's backend is different and might break its training process.

If you have a code for regression using only numpy and the like, You can post it here.
You can probably try writing your own, learn how regression works and start writing it from scratch. In fact, you can actually use the PSO NN code that I you used from the PySwarms documentation, you only have to change the loss function.

In the previous implementation, we're using the cross-entropy loss function. Try using a mean-square-error instead of cross-entropy

**Comment by JaveJao** [comment]
I will try it, sir! Thank you very much. I'll leave for school now :D Will comeback some other time!

**Comment by ljvmiranda921** [comment]
Closing this because it's not related to the actual PySwarms module itself
